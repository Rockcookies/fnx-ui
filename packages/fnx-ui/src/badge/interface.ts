import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
