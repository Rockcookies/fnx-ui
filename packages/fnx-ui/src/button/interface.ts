import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
