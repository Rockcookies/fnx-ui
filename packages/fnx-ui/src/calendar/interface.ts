import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
