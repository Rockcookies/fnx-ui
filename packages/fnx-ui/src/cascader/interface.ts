import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
