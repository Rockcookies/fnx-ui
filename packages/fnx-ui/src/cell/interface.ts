import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
