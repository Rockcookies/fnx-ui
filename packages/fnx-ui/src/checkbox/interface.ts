import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
