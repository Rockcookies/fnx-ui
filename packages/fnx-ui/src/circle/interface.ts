import { HTMLAttributes } from 'react';

export type CircleLinecap = 'butt' | 'round' | 'square';

export type CircleGapPosition = 'top' | 'right' | 'bottom' | 'left';

export type CircleStringGradients = { [percentage: string]: string };

export interface CircleComponentProps {
	size?: string | number;
	strokeColor?: string | CircleStringGradients;
	strokeLinecap?: CircleLinecap;
	strokeWidth?: string | number;
	trailColor?: string;
	progress?: number;
	clockwise?: boolean;
	speed?: number;
	fill?: string;
	gapDegree?: number;
	gapPosition?: CircleGapPosition;
}

export type CircleProps = CircleComponentProps & HTMLAttributes<HTMLDivElement>;
