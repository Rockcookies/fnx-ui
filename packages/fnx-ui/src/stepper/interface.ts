import { HTMLAttributes } from 'react';

export type StepperValue = string | number;

export type StepperMode = 'string' | 'number';

export interface StepperComponentProps<T extends StepperValue = StepperValue> {
	value?: T | null;
	defaultValue?: T | null;
	onChange?: (value: T | null) => void;
	mode?: StepperMode;
	min?: T;
	max?: T;
	allowEmpty?: boolean;
	step?: StepperValue;
	precision?: number;
	buttonSize?: number | string;
	inputWidth?: number | string;
	showInput?: boolean;
	showPlus?: boolean;
	showMinus?: boolean;
	disabled?: boolean;
	disablePlus?: boolean;
	disableMinus?: boolean;
	disableInput?: boolean;
	longPress?: boolean;
	onStep?: (
		value: T,
		info: { offset: StepperValue; type: 'minus' | 'plus' },
	) => void;
}

export type StepperProps<T extends StepperValue = StepperValue> =
	StepperComponentProps<T> &
		Omit<
			HTMLAttributes<HTMLSpanElement>,
			'value' | 'defaultValue' | 'onInput' | 'onChange'
		>;

export interface StepperRef {
	root: HTMLSpanElement | null;
	input: HTMLInputElement | null;
}
