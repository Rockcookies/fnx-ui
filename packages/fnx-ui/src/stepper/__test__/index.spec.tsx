import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import TestsDOM from '../../../test/dom';
import Stepper from '../index';
import { StepperProps } from '../interface';

describe('<Stepper/>', () => {
	const renderFn = (props: StepperProps) => {
		const { container, rerender } = render(<Stepper {...props} />);

		const minusBtn = TestsDOM.mustQuerySelector(
			container,
			'.fnx-stepper__minus',
		);
		const plusBtn = TestsDOM.mustQuerySelector(
			container,
			'.fnx-stepper__plus',
		);

		const input = TestsDOM.mustQuerySelector<HTMLInputElement>(
			container,
			'.fnx-stepper__input',
		);

		return {
			container,
			rerender,
			minusBtn,
			plusBtn,
			input,
			changeValue: (value: string) => {
				act(() => {
					fireEvent.focus(input);
					jest.advanceTimersByTime(1000);
				});

				fireEvent.change(input, { target: { value } });

				act(() => {
					fireEvent.blur(input);
					jest.advanceTimersByTime(1000);
				});
			},
		};
	};

	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('should disable buttons and input when using disabled prop', () => {
		const { container } = renderFn({
			disabled: true,
		});

		expect(container).toMatchSnapshot();
	});

	it('should make input readonly when using disable-input prop', () => {
		const { input } = renderFn({ disableInput: true });

		expect(input.getAttribute('readonly')).not.toBeNull();
	});

	it('should emit onStep event when clicking the minus/plus button', () => {
		const onStep = jest.fn();

		const { minusBtn, plusBtn, rerender } = renderFn({
			defaultValue: 3,
			onStep,
			disableMinus: true,
			disablePlus: true,
		});

		fireEvent.click(minusBtn);
		expect(onStep).not.toHaveBeenCalled();

		fireEvent.click(plusBtn);
		expect(onStep).not.toHaveBeenCalled();

		rerender(<Stepper defaultValue={3} onStep={onStep}></Stepper>);

		fireEvent.click(minusBtn);
		expect(onStep).toHaveBeenCalledWith('2', { offset: -1, type: 'minus' });

		fireEvent.click(plusBtn);
		expect(onStep).toHaveBeenCalledWith('3', { offset: 1, type: 'plus' });
	});

	test('should disable plus button when disable-plus prop is true', async () => {
		const { plusBtn, rerender } = renderFn({
			disablePlus: false,
		});

		expect(plusBtn.classList).not.toContain('fnx-stepper__plus--disabled');

		rerender(<Stepper disablePlus />);

		expect(plusBtn.classList).toContain('fnx-stepper__plus--disabled');
	});

	test('should disable minus button when disable-minus prop is true', async () => {
		const { minusBtn, rerender } = renderFn({
			disableMinus: false,
		});

		expect(minusBtn.classList).not.toContain(
			'fnx-stepper__minus--disabled',
		);

		rerender(<Stepper disableMinus />);

		expect(minusBtn.classList).toContain('fnx-stepper__minus--disabled');
	});

	test('should limit max value when using max prop', async () => {
		const onChange = jest.fn();

		const { plusBtn, minusBtn } = renderFn({
			max: 2,
			defaultValue: 1,
			onChange,
			mode: 'number',
		});

		fireEvent.click(plusBtn);
		fireEvent.click(plusBtn);
		fireEvent.click(minusBtn);
		fireEvent.click(minusBtn);

		expect(onChange).toHaveBeenNthCalledWith(1, 2);
		expect(onChange).toHaveBeenNthCalledWith(2, 1);
		expect(onChange).toHaveBeenNthCalledWith(3, 0);
	});

	test('should update value after long pressing', async () => {
		const onChange = jest.fn();

		const { minusBtn, plusBtn } = renderFn({
			defaultValue: 3,
			onChange,
		});

		fireEvent.touchStart(plusBtn);

		act(() => {
			jest.advanceTimersByTime(1600);
		});

		fireEvent.touchEnd(plusBtn);

		expect(onChange).toHaveBeenLastCalledWith('4');

		fireEvent.touchStart(minusBtn);
		act(() => {
			jest.advanceTimersByTime(1200);
		});
		fireEvent.touchEnd(minusBtn);

		expect(onChange).toHaveBeenLastCalledWith('3');
	});

	test('should allow to disable long press', async () => {
		const onChange = jest.fn();

		const { minusBtn, plusBtn } = renderFn({
			onChange,
			longPress: false,
		});

		fireEvent.touchStart(plusBtn);
		act(() => {
			jest.advanceTimersByTime(1000);
			fireEvent.touchEnd(plusBtn);
		});

		fireEvent.touchStart(minusBtn);
		act(() => {
			jest.advanceTimersByTime(1200);
			fireEvent.touchEnd(minusBtn);
		});

		expect(onChange).toHaveBeenCalledTimes(0);
	});

	it('should format it after on blur', async () => {
		const onChange = jest.fn();

		const { changeValue } = renderFn({
			onChange,
			max: 5,
		});

		changeValue('44');

		expect(onChange).toHaveBeenLastCalledWith('5');
	});

	test('should format value to default value when input is invalid', async () => {
		const onChange = jest.fn();

		const { changeValue } = renderFn({
			onChange,
		});

		changeValue('');

		expect(onChange).toHaveBeenLastCalledWith('0');
	});

	it('only allow integer and set mode number', () => {
		const onChange = jest.fn();

		const { changeValue } = renderFn({
			onChange: onChange,
			precision: 0,
			mode: 'number',
		});

		changeValue('4');
		expect(onChange).toHaveBeenLastCalledWith(4);

		changeValue('1.2');
		expect(onChange).toHaveBeenLastCalledWith(1);
	});

	it('precision prop', () => {
		const onChange = jest.fn();
		const { plusBtn } = renderFn({
			step: 0.2,
			precision: 2,
			onChange: onChange,
		});

		fireEvent.click(plusBtn);
		expect(onChange).toHaveBeenLastCalledWith('0.20');
	});

	it('input-width prop', () => {
		const { container } = renderFn({
			inputWidth: '80px',
		});
		expect(container).toMatchSnapshot();
	});

	it('button-size prop', () => {
		const { container } = renderFn({
			buttonSize: '15px',
		});
		expect(container).toMatchSnapshot();
	});

	it('min value is 3', () => {
		const onChange = jest.fn();
		const { changeValue } = renderFn({
			min: 3,
			onChange: onChange,
		});

		changeValue('1');

		expect(onChange).toHaveBeenLastCalledWith('3');
	});

	it('show-plus & show-minus props', () => {
		const { container } = render(
			<Stepper showPlus={false} showMinus={false}></Stepper>,
		);

		expect(container).toMatchSnapshot();
	});

	it('allow-empty prop', () => {
		const onChange = jest.fn();
		const { changeValue } = renderFn({
			allowEmpty: true,
			onChange: onChange,
		});

		changeValue('');

		expect(onChange).toHaveBeenLastCalledWith(null);
	});
});
