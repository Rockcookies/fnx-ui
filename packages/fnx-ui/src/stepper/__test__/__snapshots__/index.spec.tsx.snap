// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper/> button-size prop 1`] = `
<div>
  <span
    class="fnx-stepper"
    role="group"
  >
    <button
      class="fnx-stepper__minus"
      style="width: 15px; height: 15px;"
      type="button"
    />
    <input
      aria-valuenow="0"
      class="fnx-stepper__input"
      inputmode="decimal"
      role="spinbutton"
      style="height: 15px;"
      type="text"
      value="0"
    />
    <button
      class="fnx-stepper__plus"
      style="width: 15px; height: 15px;"
      type="button"
    />
  </span>
</div>
`;

exports[`<Stepper/> input-width prop 1`] = `
<div>
  <span
    class="fnx-stepper"
    role="group"
  >
    <button
      class="fnx-stepper__minus"
      type="button"
    />
    <input
      aria-valuenow="0"
      class="fnx-stepper__input"
      inputmode="decimal"
      role="spinbutton"
      style="width: 80px;"
      type="text"
      value="0"
    />
    <button
      class="fnx-stepper__plus"
      type="button"
    />
  </span>
</div>
`;

exports[`<Stepper/> should disable buttons and input when using disabled prop 1`] = `
<div>
  <span
    class="fnx-stepper"
    role="group"
  >
    <button
      aria-disabled="true"
      class="fnx-stepper__minus fnx-stepper__minus--disabled"
      type="button"
    />
    <input
      aria-valuenow="0"
      class="fnx-stepper__input"
      disabled=""
      inputmode="decimal"
      role="spinbutton"
      type="text"
      value="0"
    />
    <button
      aria-disabled="true"
      class="fnx-stepper__plus fnx-stepper__plus--disabled"
      type="button"
    />
  </span>
</div>
`;

exports[`<Stepper/> show-plus & show-minus props 1`] = `
<div>
  <span
    class="fnx-stepper"
    role="group"
  >
    <button
      class="fnx-stepper__minus fnx-stepper__minus--hidden"
      type="button"
    />
    <input
      aria-valuenow="0"
      class="fnx-stepper__input"
      inputmode="decimal"
      role="spinbutton"
      type="text"
      value="0"
    />
    <button
      class="fnx-stepper__plus fnx-stepper__plus--hidden"
      type="button"
    />
  </span>
</div>
`;
